source/agora/test/Base.d

```diff
@@ -559,55 +564,98 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
     {
-        BanManager.Config ban_conf =
+        auto other_nodes =
+            node_confs
+                .filter!(conf => conf != self)
+                .map!(conf => conf.key_pair.address.toString());
+
+        auto quorum_keys =
+            node_confs
+                .filter!(conf => conf.is_validator)
+                .map!(conf => conf.key_pair.address).array.assumeUnique;
+
+        Config conf =
         {
-            max_failed_requests : max_failed_requests,
-            ban_duration: 300
+            banman : ban_conf,
+            node : self,
+            network : configure_network ? assumeUnique(other_nodes.array) : null,
+            quorum : { nodes : quorum_keys }
         };
 
+        return conf;
+    }
+
+    Config makeCyclicConfig (size_t idx, NodeConfig self, NodeConfig[] node_confs)
+    {
+        auto prev_idx = idx == 0 ? node_confs.length - 1 : idx - 1;
+
         auto other_nodes =
             node_confs
                 .filter!(conf => conf != self)
                 .map!(conf => conf.key_pair.address.toString());
 
-        auto quorum_keys = assumeUnique(
-            node_confs.map!(conf => conf.key_pair.address).array);
+        immutable quorum_keys = [self.key_pair.address, node_confs[prev_idx].key_pair.address];
 
         Config conf =
         {
             banman : ban_conf,
             node : self,
             network : configure_network ? assumeUnique(other_nodes.array) : null,
-            quorum : { nodes : quorum_keys }
+            quorum : { nodes : quorum_keys /*, threshold : 2*/ }  // fails with 2
         };
 
         return conf;
     }
 
+    NodeConfig[] node_configs;
+    Config[] configs;
+
     final switch (topology)
     {
     case NetworkTopology.Simple:
-        auto node_configs = iota(nodes).map!(_ => makeNodeConfig()).array;
-        auto configs = iota(nodes)
+        node_configs = iota(nodes).map!(_ => makeNodeConfig()).array;
+        configs = iota(nodes)
             .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
+        break;
 
-        auto net = new APIManager();
-        foreach (idx, ref conf; configs)
-        {
-            const address = node_configs[idx].key_pair.address;
-            net.createNewNode(address, conf);
-        }
+    case NetworkTopology.OneNonValidator:
+        node_configs ~= iota(nodes).map!(_ => makeNodeConfig()).array;
+        node_configs[$ - 1].is_validator = false;
+        configs = iota(nodes)
+            .map!(idx => makeConfig(node_configs[idx], node_configs)).array;
+        break;
 
-        return net;
+    case NetworkTopology.Cyclic:
+        node_configs = iota(nodes).map!(_ => makeNodeConfig()).array;
+        node_configs.each!((ref conf) => conf.min_listeners = 1);
+        configs = iota(nodes)
+            .map!(idx => makeCyclicConfig(idx, node_configs[idx], node_configs))
+                .array;
 
+        break;
+
+    case NetworkTopology.SinglePoF:
```

Comment:
Why the move ?

Reply:
The diff seems to make it harder to see, but previously we only created nodes for the `NetworkTopology.Simple` config, now we create configs for `NetworkTopology.Simple`, `NetworkTopology.OneNonValidator`, and `NetworkTopology.Cyclic`, and we move the instantiation of the nodes after the configs were built (so we don't duplicate node creation code).

Reply:
I meant that L608 (`case NetworkTopology.SinglePoF:`) has just been reordered and is still doing the same (assert(0)), but not a big deal.

Reply:
Oh good eye! I'll fix that up.